brontide/read_buffer_pool.go

```diff
@@ -0,0 +1,85 @@
+package brontide
+
+import (
+	"time"
+
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/queue"
+)
+
+const (
+	// DefaultGCInterval is the default interval that the WriteBufferPool
+	// will perform a sweep to see which expired buffers can be released to
+	// the runtime.
+	DefaultGCInterval = 15 * time.Second
+
+	// DefaultExpiryInterval is the default, minimum interval that must
+	// elapse before a WriteBuffer will be released. The maximum time before
+	// the buffer can be released is equal to the expiry interval plus the
+	// gc interval.
+	DefaultExpiryInterval = 30 * time.Second
+
+	// readBufferSize represents the size of the maximum message that can be
+	// read off  the wire by brontide. The buffer is used to hold the
+	// ciphertext while the  brontide state machine decrypts the message.
+	readBufferSize = lnwire.MaxMessagePayload + macSize
+)
+
+// ReadBuffer is a static byte array sized to the maximum-allowed message size,
+// plus 16 bytes for the MAC.
+type ReadBuffer [readBufferSize]byte
+
+// Recycle zeroes the ReadBuffer, making it fresh for another use.
+// Zeroing the buffer using a logarithmic number of calls to the optimized copy
+// method.  Benchmarking shows this to be ~30 times faster than a for loop that
+// sets each index to 0 for this buffer size. Inspired by:
+// https://stackoverflow.com/questions/30614165/is-there-analog-of-memset-in-go
+func (b *ReadBuffer) Recycle() {
+	b[0] = 0
+	for i := 1; i < readBufferSize; i *= 2 {
+		copy(b[i:], b[:i])
+	}
+}
+
+// newRecyclableReadBuffer is a constructor that returns a ReadBuffer as a
+// queue.Recycler.
+func newRecyclableReadBuffer() queue.Recycler {
```

Comment:
Is it necessary to return the interface rather than the struct? If you return the struct/type, then it case be passed into all contexts a  `queue.Recycler` is accepted and also where a regular slice is accepted (once you slice it). As is now, callers will have to type assert it back to a plain buffer.

Reply:
the gcqueue has to accept some generic `New()` method, the only other thing that makes sense is a `func() interface{}`, otherwise we'd have to build a unique gcqueue for each distinct return type.

In this `func() interface{}` case, then the gcqueue has to perform an expensive interface -> interface conversion to make it a `queue.Recycler`. So there is a performance argument to having it return the `queue.Recycler`directly. Type asserting from interface to element is more efficient so I opted for this

The `queue.Recycler` can also be used for things that are not byte arrays, so idt we want to assume they can always be sliced